File: 240929/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
n, m = map(int, input().split())
area = []
for i in range(n):
    area.append(list(map(int, input().split())))

answer = []
for i in range(n):
    answer.append([0] * m)

visited = []
for i in range(n):
    visited.append([0] * m)

x, y = 0, 0
visited[0][0] = 1


def in_range(x, y):
    if 0 <= x < n and 0 <= y < m:
        return True
    return False

def can_go(x, y):
    if in_range(x, y) and area[x][y] == 1 and not visited[x][y]:
        return True
    return False


def dfs(x, y):
    global ans

    if x == n - 1 and y == m - 1:
        ans = True
    
    dx, dy = [0, 1], [1, 0]
    for i in range(2):
        nx, ny = x + dx[i], y + dy[i]
        if can_go(nx, ny):
            visited[nx][ny] = 1
            dfs(nx, ny)
    
    return False

ans = False
dfs(0, 0)

if ans:
    print(1)
else:
    print(0)

File: 250222/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
using namespace std;

int n, m;
int area[100][100];
bool ans = false;
// 오른쪽, 아래
int dx[] = { 0, 1 };
int dy[] = { 1, 0 };
bool visited[100][100] = {false,};

bool InRange(int x, int y) {
    return (0 <= x && x < n && 0 <= y && y < m);
}

bool CanGo(int x, int y) {
    return (InRange(x, y) && area[x][y] == 1 && !visited[x][y]);
}

void DFS(int x, int y) {
    if (x == n - 1 && y == m - 1) {
        ans = true;
        return;
    }

    for (int dr = 0; dr < 2; dr++) {
        int nx = x + dx[dr];
        int ny = y + dy[dr];

        if (CanGo(nx, ny)) {
            visited[nx][ny] = true;
            DFS(nx, ny);
        }
    }
}

int main() {
    cin >> n >> m;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            cin >> area[i][j];

    visited[0][0] = true;
    DFS(0, 0);

    cout << ans;

    return 0;
}